warehouse/accounts/auth_policy.py

```diff
@@ -10,15 +10,110 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import struct
+
+from pymacaroons import Macaroon, Verifier
+from pymacaroons.caveat_delegates import ThirdPartyCaveatVerifierDelegate
+from pymacaroons.exceptions import MacaroonException
+
 from pyramid.authentication import (
     BasicAuthAuthenticationPolicy as _BasicAuthAuthenticationPolicy,
+    CallbackAuthenticationPolicy as _CallbackAuthenticationPolicy,
     SessionAuthenticationPolicy as _SessionAuthenticationPolicy,
 )
 
 from warehouse.accounts.interfaces import IUserService
 from warehouse.cache.http import add_vary_callback
 
 
+class AccountTokenAuthenticationPolicy(_CallbackAuthenticationPolicy):
+    def __init__(self, authenticate):
+        self._authenticate = authenticate
+        self.callback = self._auth_callback
+
+        self._routes_allowed = ["forklift.legacy.file_upload"]
+
+    def unauthenticated_userid(self, request):
+        if request.matched_route.name not in self._routes_allowed:
+            return None
+
+        account_token = request.params.get("account_token")
+
+        if account_token is None:
+            return None
+
+        try:
+            macaroon = Macaroon.deserialize(account_token)
+
+            # First, check identifier and location
+            if macaroon.identifier != request.registry.settings["account_token.id"]:
+                return None
+
+            if macaroon.location != "pypi.org":
```

Comment:
You're right! According to https://ai.google/research/pubs/pub41892 this field is optional and (I just learned) even left out of the integrity checks. I'll remove this.

I've been learning more about macaroons since I first heard about them at PyCon 2018...